### PR TITLE
refactor(protocol-designer): prepare Flow Rate field for batch edit mode

### LINKED ISSUE #7206: refactor(protocol-designer): prepare Flow Rate field + modal for batch edit form
# Overview

The existing Flow Rate field components need to be refactored so that it can be reused in the Batch Edit Transfer form, while continuing to be used in single-edit mode (in SourceDestFields + MixForm).

The idea would be that for ALL the modal fields, we could take the approach explained in this ticket.
1. We factor out the stuff that varies between single-edit and batch-edit modes so that for every field "Spam" we would make new containers `SpamSingleEditField + SpamBatchEditField` which inject data like "relevant selected PipetteEntity for this field", "value for this field", "disabled state of this field".
2. `SpamSingleEditField` uses `getUnsavedForm` selector and `getDisabledFields(unsavedFormData)` util for this stuff, and "update value" is accomplished by dispatching UPDATE_FORM_FIELD. It doesn't have to handle indeterminate-value state.
3. In contrast, `SpamBatchEditField` relies on the `BatchEditForm` which holds any changed field values in its component state to pass in values, along with the new `getMultiselectFieldValues` selector. It has no connection to `getUnsavedForm`. It doesn't dispatch anything, it calls a callback passed in from BatchEditForm to indirectly update `BatchEditForm`'s component state. It does have to handle indeterminate-value state (and that's unique to batch edit)

tl;dr of the above: the refactoring pattern for modal-fields is
- SpamField: the field that also controls its modal, a dumb component (might have its own internal component state, but doesn't use useSelector or useDispatch, relies on its props being fed it for that stuff)
- SpamSingleEditField: a container of SpamField specific to single-edit mode
- SpamBatchEditField: a container of SpamField specific to batch-edit mode

# Acceptance Criteria

- [ ] Decouple `FlowRateField` from the `getUnsavedForm` selector. To `FlowRateField` component's props, add new props `pipetteEntity: PipetteEntity | null, disabled: boolean, formFlowRate: number | null`. Remove the now-unneeded prop `pipetteFieldName` (we would use the new `pipetteEntity` prop directly, which `pipetteFieldName` just).
  - If convenient, we could make containers `FlowRateFieldSingleForm` (consumed by SourceDestFields + MixForm) and separately `FlowRateFieldBatchEditForm` that populate these props.
  - At this point, it might make sense to merge `FlowRateInput` and `FlowRateField` into one functional component (both are now class-based components).
- [ ] Also, decouple dispatching `changeFormInput` action from the `updateValue` prop. It should be up to `FlowRateFieldSingleForm` to make `updateValue` prop dispatch that action. And `FlowRateFieldBatchEditForm` will not wire up `updateValue` but just itself have the same prop and pass it down (the batch edit form component itself would pass down a fn to do a setState on the batch edit form's stateful field data). In other words, the DTP of `FlowRateField` should be removed, it's the responsibility of `FlowRateFieldSingleForm / FlowRateFieldBatchEditForm`  to hook up `updateValue`
- [ ] Make sure all existing functionality is preserved around the Flow Rate field in single-edit forms (both Transfer and Mix). Especially watch out for stale data issues (outlined below)
- [ ] For `FlowRateFieldBatchEditForm` only, allow it to display an indeterminate state `-` via the `placeholder` attr. Might involve new props
- [ ] It's up to the assignee of this ticket whether to include fully implementing `FlowRateFieldBatchEditForm` in this ticket. It is likely will be too much for one PR, and is likely to get blocked by other tickets for selectors etc. If so, make a placeholder ticket for that work and add it to the epic. If that's the case, `FlowRateFieldBatchEditForm` can be left as a skeleton, perhaps with hard-coded values for the props

### `FlowRateInput`

- Could be converted to functional component w hooks.
- Initial state is derived from props
- Pristinity is boolean state (whole form is pristine or not)
- Props come from FlowRateField
- Radio button state is managed by a boolean in the component state (though radio button UI is a text value)

> When flow rate is falsey (including 0), it means 'use default'

NOT just `null`, but any falsey value!

### FlowRateField

- A Redux-connected container of `FlowRateInput`.
- It uses a `key` prop so if name or value of the field changes, it forces a re-render of the modal component (thereby ensuring a fresh initial state, no stale state carried over)
- Sets `defaultFlowRate` in STP (from other selectors)
- Uses `getDisabledFields` to determine `disabled` prop value
- Min flow rate always `0`, max always from pipette spec (in weird case where there is no valid selected pipette, max is `Infinity`)
- DTP: `updateValue` prop dispatches `changeFormInput` (`updateValue` prop is called in `FlowRateInput` when user clicks modal's save button)

### What is missing for batch edit version

The `FlowRateField` container uses `getUnsavedForm` directly, batch edit doesn't use that. It needs to get both the flow rate value and the pipette ID from component state in the case of batch-edit, and from `getUnsavedForm` in the case of single-edit.

We def want to reuse the way it gets defaults from the pipette

Also, we can't dispatch an action. That `updateValue` prop needs to be just passed in, remove the DTP

Is there a cleaner way to reset the state when props change? Also let's try converting both components into functional ones that use hooks.

### PR DESCRIPTION
**Pull request by IanLondon**
# Overview

Closes #7206 though the implementation is different since now we have the `FieldProps` interface and we have `formData` available in the form.

Addresses part of #7301

# Changelog

Remove all dependencies from Flow Rate field that are bound to single-edit mode. The component is "smart" enough to get the pipette spec it needs, but without `getUnsavedForm` or dispatching `'CHANGE_FORM_INPUT'` on its own. Instead, it relies on getting the form data and the updateValue handler passed down into it.

# Review requests

- [ ] Code review. Are there any remaining dependencies on single-edit mode?
- [ ] The flow rate fields in Transfer and Mix forms should work the same: min/max should match pipette (though min is always 0.1), validation should work the same, saving the modal should update the form field. Compare to prod side-by-side.
- [ ] Once again, discovered some missing fields in the mix form `getDefaultsForStepType` :grimacing:. This SHOULD change a buggy behavior where changing the pipette in the mix form may not have reset the flow rates. This should not (AFAIK... and please think this through) cause any kind of problems where importing protocols would change the flow rates.

# Risk assessment

Medium, messing with Flow Rate field
Medium-high, adding missing fields to getDefaultsForStepType

### CODE REVIEW

File: protocol-designer/src/steplist/formLevel/handleFormChange/test/mix.test.js [hunk 1]
```diff
@@ -57,12 +57,22 @@ describe('well selection should update', () => {
 
   it('pipette cleared', () => {
     const patch = { pipette: null }
-    expect(handleFormHelper(patch, form)).toEqual({ ...patch, wells: [] })
+    expect(handleFormHelper(patch, form)).toEqual({
```
Comment (IanLondon) [inline protocol-designer/src/steplist/formLevel/handleFormChange/test/mix.test.js:60]: I am pretty sure this is a bug fix for a bug we never noticed. Looking at mix's handleFormChange fn, it was supposed to set the flow rates to null when pipette is changed, but it expected those fields to come from getDefaultsForStepType so then it never actually worked

File: protocol-designer/src/components/StepEditForm/fields/FlowRateField/FlowRateInput.js [hunk 2]
```diff
@@ -37,18 +36,18 @@ type State = {
...
+export class FlowRateInput extends React.Component<FlowRateInputProps, State> {
+  constructor(props: FlowRateInputProps) {
     super(props)
     this.state = this.getStateFromProps(props)
   }
 
-  getStateFromProps: (props: Props) => State = props => {
-    const { formFlowRate } = props
+  getStateFromProps: (props: FlowRateInputProps) => State = props => {
+    const { value } = props
```
Comment (IanLondon) [inline protocol-designer/src/components/StepEditForm/fields/FlowRateField/FlowRateInput.js:46]: the prop `formFlowRate` was just a more confusing name for `value`

**Review by shlokamin — APPROVED** [review]
LGTM, and it seems to work the exact same as on prod 👍🏽

**Review by Kadee80 — APPROVED** [review]
🏄
